<h3>Industry-Based Hands-On Exercises</h3>
<p><b>PSG: NumPy Mathematical Operations, Equations & Sequences</b></p>

<hr>

<h4>HANDS-ON EXERCISE 1</h4>
<p><b>Business Analytics — Sales Performance Calculator</b></p>

<p><b>Scenario</b><br>
Imagine you are supporting the analytics team of a growing retail business.<br>
The business recorded the following daily sales for one week:</p>

<pre><code>daily_sales = np.array([125000, 150000, 175000, 140000, 190000, 210000, 160000])</code></pre>

<p>The management team wants to understand the week's performance.<br>
Your task is to use NumPy to perform basic numerical analysis on the sales data.</p>

<br>

<p><b>Your Tasks</b></p>

<p><b>Task 1 — Create the Dataset</b></p>

<p>Import NumPy:</p>
<pre><code>import numpy as np</code></pre>

<p>Create the <code>daily_sales</code> array using the values provided above.</p>

<p><b>Print:</b></p>
<ul>
  <li>The array</li>
  <li>The type of the array</li>
  <li>The shape of the array</li>
</ul>

In [30]:
import numpy as np

daily_sales = np.array([125000, 150000, 175000, 140000, 190000, 210000, 160000])

print(daily_sales)
print(type(daily_sales))
print(daily_sales.shape)

[125000 150000 175000 140000 190000 210000 160000]
<class 'numpy.ndarray'>
(7,)


<p><b>Task 2 — Calculate Total Sales</b><br>
Use NumPy to calculate the total sales for the week.</p>

<p>You should use:</p>
<pre><code>np.sum()</code></pre>

<p>Store the result in a meaningful variable.<br>
For example:</p>
<pre><code>total_sales = ...</code></pre>

<p>Then print the result.</p>

In [31]:
total_sales = np.sum(daily_sales)

print(f"Total Sales: {total_sales:,}")

Total Sales: 1,150,000


<p><b>Task 3 — Calculate Average Daily Sales</b><br>
Use NumPy to calculate the average daily sales.</p>

<p>You should use:</p>
<pre><code>np.mean()</code></pre>

<p>Store the result as:</p>
<pre><code>average_sales = ...</code></pre>

<p>Print the result.</p>

In [32]:
average_sales = np.mean(daily_sales)

print(f"Average Sales: {average_sales:,.2f}")

Average Sales: 164,285.71


<p><b>Task 4 — Apply a Sales Adjustment</b><br>
Suppose the company wants to estimate sales after a hypothetical 10% increase.</p>

<p>Create a new array representing the adjusted sales.</p>

<p><b>Instruction:</b><br>
Do not manually change each value.<br>
Use NumPy's element-wise operations.</p>

<p><b>Buddy Hint:</b><br>
Think about how you could multiply the entire array by a factor (e.g., <code>1.10</code>) to scale all daily figures simultaneously.</p>

In [33]:
adjusted_sales = daily_sales * 1.10

print(adjusted_sales)

[137500. 165000. 192500. 154000. 209000. 231000. 176000.]


In [34]:
""

''

<p><b>Task 5 — Compare Original and Adjusted Sales</b><br>
Create an array showing the difference between:</p>

<pre><code>Adjusted sales − Original sales</code></pre>

<p><b>Instruction:</b><br>
This will show how much additional revenue each day represents under the hypothetical 10% increase.</p>

<p><b>Buddy Hint:</b><br>
Perform a direct array subtraction (<code>adjusted_sales - daily_sales</code>) to leverage NumPy's vectorized element-wise arithmetic.</p>

In [35]:
sales_difference = adjusted_sales - daily_sales

print(sales_difference)

[12500. 15000. 17500. 14000. 19000. 21000. 16000.]


<p><b>Task 6 — Use NumPy's Mathematical Functions</b><br>
Use an appropriate NumPy function to calculate the average daily sales.</p>

<p>Then compare your result with the total sales.</p>

<br>

<p><b>Write a short Markdown explanation answering:</b><br>
What does the average daily sales figure tell the business that the total sales figure does not?</p>

<br>

<p><b>Buddy Hint:</b><br>
Consider how total revenue measures overall volume, whereas average daily sales establishes a baseline metric to gauge daily performance trends and operational consistency regardless of week length.</p>

Average daily sales tells the business how it is performing on a day to day basis <br>
Total sales on the other hand, gives insight into the gross sales made over a longer period of time usually weekly, <br>
monthly, quarterly or annually. This gives the business a proper understanding of overall volume.

In [36]:
sales_minus_average = total_sales - average_sales

print(f"Difference between total and average sales: {sales_minus_average:,.2f}")

Difference between total and average sales: 985,714.29


<h4>HANDS-ON EXERCISE 2</h4>
<p><b>Education Analytics — Student Performance Analysis</b></p>

<p><b>Scenario</b><br>
You are assisting an education analytics team.<br>
A training program recorded the scores of a group of students:</p>

<pre><code>scores = np.array([62, 75, 81, 69, 88, 94, 73, 85, 77, 91])</code></pre>

<p>The program coordinator wants to understand the overall performance of the students and how individual scores differ from the group average.<br>
Your job is to perform the numerical analysis using NumPy.</p>

<br>

<p><b>Your Tasks</b></p>

<p><b>Task 1 — Create and Inspect the Data</b></p>

<p>Create the NumPy array:</p>
<pre><code>scores</code></pre>

<p><b>Print:</b></p>
<ul>
  <li>The scores</li>
  <li>The type</li>
  <li>The shape</li>
</ul>

In [37]:
scores = np.array([62, 75, 81, 69, 88, 94, 73, 85, 77, 91])

print(scores)
print(type(scores))
print(scores.shape)

[62 75 81 69 88 94 73 85 77 91]
<class 'numpy.ndarray'>
(10,)


<p><b>Task 2 — Calculate the Mean Score</b><br>
Calculate the average score using NumPy.</p>

<p>Store the result as:</p>
<pre><code>mean_score = ...</code></pre>

<p>Print the result.</p>

In [38]:
mean_score = np.mean(scores)

print(f"Mean Score: {mean_score}")

Mean Score: 79.5


<p><b>Task 3 — Calculate Each Student's Deviation</b><br>
Now determine how far each student's score is from the average.</p>

<p>Use:</p>
<pre><code>score − mean score</code></pre>

<p>Your result should be a NumPy array.<br>
Store it as:</p>
<pre><code>deviations = ...</code></pre>

<p>Print the deviations.</p>

In [39]:
deviations = scores - mean_score

print(deviations)

[-17.5  -4.5   1.5 -10.5   8.5  14.5  -6.5   5.5  -2.5  11.5]


<p><b>Task 4 — Identify Positive and Negative Deviations</b><br>
Inspect the deviation array.</p>

<p><b>Explain:</b></p>
<ul>
  <li>What does a positive deviation mean?</li>
  <li>What does a negative deviation mean?</li>
  <li>What does a value close to zero mean?</li>
</ul>

<p>Write your explanation in Markdown.</p>

A positive deviation means that the score is greater than the mean score. <br>
Conversely, a negative deviation indicates a score smaller than the mean score. <br>
A deviation of zero or close to zero means that the score is the mean value of the scores

<p><b>Task 5 — Square the Deviations</b><br>
Create a new array containing the squared deviations.</p>

<p>For example:</p>
<pre><code>squared_deviations = deviations ** 2</code></pre>

<p>Print the result.</p>

<p>Explain why squaring the deviations changes the negative values into positive values.</p>

In [40]:
squared_deviations = deviations ** 2

print(squared_deviations)

[306.25  20.25   2.25 110.25  72.25 210.25  42.25  30.25   6.25 132.25]


<p><b>Task 6 — Calculate the Sum of Squared Deviations</b><br>
Use:</p>

<pre><code>np.sum()</code></pre>

<p>to calculate the total of the squared deviations.</p>

<p>Store it as:</p>
<pre><code>sum_squared_deviations = ...</code></pre>

<p>Print the result.</p>

In [41]:
sum_squared_deviations = np.sum(squared_deviations)

print(sum_squared_deviations)

932.5


<p><b>Task 7 — Connect the Steps to Standard Deviation</b><br>
Using the process discussed in the lesson, identify the stages involved in calculating standard deviation.</p>

<p>Your Markdown section should show the process conceptually:</p>
<ol>
  <li>Calculate the mean</li>
  <li>Calculate deviations</li>
  <li>Square the deviations</li>
  <li>Sum the squared deviations</li>
  <li>Continue with the appropriate division</li>
  <li>Take the square root</li>
</ol>

<p>Then explain in your own words why standard deviation is useful when analyzing student scores.</p>

The stages involved in calculating standard deviation are as follows:
<ol>
  <li>Calculate the mean</li>
  <li>Calculate deviations</li>
  <li>Square the deviations</li>
  <li>Sum the squared deviations</li>
  <li>Compute the sample size</li>
  <li>Calculate sum_of_squared_deviations / (sample_size - 1)</li>
  <li>Take the square root</li>
</ol>

Standard deviation is useful in the analysis of student scores because it enables us to know how close the individual student scores are to the average score. This gives an insight on overall student performance. A small standard deviation means the individual scores are close to the mean score. The opposite is true for a high standard deviation


<h4>HANDS-ON EXERCISE 3</h4>
<p><b>Engineering &amp; Scientific Computing — Trigonometric Series and Convergence</b></p>

<p><b>Scenario</b><br>
You are supporting an engineering/scientific computing team working with a mathematical series.<br>
The team wants to understand how an approximation changes as more terms are included in a calculation.<br>
You will use NumPy to create a sequence of values, work with a trigonometric function, and observe the idea of convergence.</p>

<br>

<p><b>Your Tasks</b></p>

<p><b>Task 1 — Define the Angle</b></p>

<p>Create:</p>
<pre><code>theta = 30</code></pre>

<p>This represents an angle in degrees.<br>
Convert the angle to radians using:</p>

<pre><code>np.radians()</code></pre>

<p>Store the converted value.<br>
For example:</p>

<pre><code>theta_radians = ...</code></pre>

<p>Print the result.</p>

In [42]:
theta = 30
theta_radians = np.radians(theta)

print(theta_radians)

0.5235987755982988


<p><b>Task 2 — Create the Sequence of Terms</b><br>
Create a sequence of values for k.</p>

<p>Start with:</p>
<pre><code>k = np.arange(1, 100, 1)</code></pre>

<p>Print the resulting array.<br>
Then inspect:</p>
<ul>
  <li>The first value</li>
  <li>The last value</li>
  <li>The number of values generated</li>
</ul>

<p>Think carefully about how the start, stop, and step values affect the sequence.</p>

In [43]:
k = np.arange(1,100)
print(k)

[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72
 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96
 97 98 99]


<p><b>Task 3 — Calculate the Trigonometric Component</b><br>
Use NumPy's sine function:</p>

<pre><code>np.sin()</code></pre>

<p>with your converted angle.<br>
Store the result.</p>

In [44]:
sin_k_theta = np.sin(k * theta_radians)

print(sin_k_theta)

[ 5.00000000e-01  8.66025404e-01  1.00000000e+00  8.66025404e-01
  5.00000000e-01  1.22464680e-16 -5.00000000e-01 -8.66025404e-01
 -1.00000000e+00 -8.66025404e-01 -5.00000000e-01 -2.44929360e-16
  5.00000000e-01  8.66025404e-01  1.00000000e+00  8.66025404e-01
  5.00000000e-01  3.67394040e-16 -5.00000000e-01 -8.66025404e-01
 -1.00000000e+00 -8.66025404e-01 -5.00000000e-01 -4.89858720e-16
  5.00000000e-01  8.66025404e-01  1.00000000e+00  8.66025404e-01
  5.00000000e-01  2.38868024e-15 -5.00000000e-01 -8.66025404e-01
 -1.00000000e+00 -8.66025404e-01 -5.00000000e-01 -7.34788079e-16
  5.00000000e-01  8.66025404e-01  1.00000000e+00  8.66025404e-01
  5.00000000e-01  8.57252759e-16 -5.00000000e-01 -8.66025404e-01
 -1.00000000e+00 -8.66025404e-01 -5.00000000e-01 -9.79717439e-16
  5.00000000e-01  8.66025404e-01  1.00000000e+00  8.66025404e-01
  5.00000000e-01  4.65489580e-15 -5.00000000e-01 -8.66025404e-01
 -1.00000000e+00 -8.66025404e-01 -5.00000000e-01 -4.77736048e-15
  5.00000000e-01  8.66025

<p><b>Task 4 — Build the Series Terms</b><br>
Using the k array, create the individual terms of the series by dividing the sine component by k.</p>

<p>Think about how NumPy allows one value to interact with every element of an array.<br>
Your calculation should follow the structure:</p>

<pre><code>sin(theta) / k</code></pre>

<p>Store the result in a meaningful variable.</p>

In [45]:
series_terms = sin_k_theta / k

print(series_terms)

[ 5.00000000e-01  4.33012702e-01  3.33333333e-01  2.16506351e-01
  1.00000000e-01  2.04107800e-17 -7.14285714e-02 -1.08253175e-01
 -1.11111111e-01 -8.66025404e-02 -4.54545455e-02 -2.04107800e-17
  3.84615385e-02  6.18589574e-02  6.66666667e-02  5.41265877e-02
  2.94117647e-02  2.04107800e-17 -2.63157895e-02 -4.33012702e-02
 -4.76190476e-02 -3.93647911e-02 -2.17391304e-02 -2.04107800e-17
  2.00000000e-02  3.33086694e-02  3.70370370e-02  3.09294787e-02
  1.72413793e-02  7.96226746e-17 -1.61290323e-02 -2.70632939e-02
 -3.03030303e-02 -2.54713354e-02 -1.42857143e-02 -2.04107800e-17
  1.35135135e-02  2.27901422e-02  2.56410256e-02  2.16506351e-02
  1.21951220e-02  2.04107800e-17 -1.16279070e-02 -1.96823955e-02
 -2.22222222e-02 -1.88266392e-02 -1.06382979e-02 -2.04107800e-17
  1.02040816e-02  1.73205081e-02  1.96078431e-02  1.66543347e-02
  9.43396226e-03  8.62017740e-17 -9.09090909e-03 -1.54647394e-02
 -1.75438596e-02 -1.49314725e-02 -8.47457627e-03 -7.96226746e-17
  8.19672131e-03  1.39681

<p><b>Task 5 — Calculate the Sum</b><br>
Use:</p>

<pre><code>np.sum()</code></pre>

<p>to calculate the sum of the generated terms.<br>
Store and print the result.</p>

In [46]:
sum_of_terms = np.sum(series_terms)

print(sum_of_terms)

1.3136674864204645


<p><b>Task 6 — Increase the Number of Terms</b><br>
Now repeat the experiment with a larger number of terms.<br>
For example, create sequences using progressively larger ranges.</p>

<p>You can investigate:</p>
<pre><code>np.arange(1, 100, 1)</code></pre>

<p>then:</p>
<pre><code>np.arange(1, 1000, 1)</code></pre>

<p>and then:</p>
<pre><code>np.arange(1, 10000, 1)</code></pre>

<p>For each version, calculate the corresponding series result.<br>
Record the results.</p>

In [47]:
# using 1,000 series

increased_k = np.arange(1, 1000, 1)

adjusted_series_sum = np.sum(np.sin(increased_k * theta_radians)/increased_k)

print(adjusted_series_sum)

1.3094937000502784


In [48]:
# using 10,000 series

increased_k = np.arange(1, 10000, 1)

adjusted_series_sum = np.sum(np.sin(increased_k * theta_radians)/increased_k)

print(adjusted_series_sum)

1.3090469066682815


<p><b>Task 7 — Investigate Convergence</b><br>
Compare the results from the different numbers of terms.</p>

<p>Create a simple table in your notebook containing:</p>

<table>
  <thead>
    <tr>
      <th>Number of Terms</th>
      <th>Series Result</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>99</td>
      <td>Your result</td>
    </tr>
    <tr>
      <td>999</td>
      <td>Your result</td>
    </tr>
    <tr>
      <td>9,999</td>
      <td>Your result</td>
    </tr>
  </tbody>
</table>

<br>

<p>Then answer:<br>
What happens to the result as the number of terms increases?</p>

In [49]:
# 99 series

increased_k = np.arange(1, 99, 1)

adjusted_series_sum = np.sum(np.sin(increased_k * theta_radians)/increased_k)

print(adjusted_series_sum)

1.3035664763194543


In [50]:
# 999 series

increased_k = np.arange(1, 999, 1)

adjusted_series_sum = np.sum(np.sin(increased_k * theta_radians)/increased_k)

print(adjusted_series_sum)

1.3084926990492773


In [51]:
# 9999 series

increased_k = np.arange(1, 9999, 1)

adjusted_series_sum = np.sum(np.sin(increased_k * theta_radians)/increased_k)

print(adjusted_series_sum)

1.3089468966672813


| Number of terms | Series result |
| --- | --- |
| 99 | 1.3035664763194543 |
| 999 | 1.3084926990492773 |
| 9999 | 1.3089468966672813 |

<p><b>Task 8 — Explain Convergence</b><br>
Write a Markdown explanation of what convergence means in the context of your experiment.</p>

<p><b>Your explanation should answer:</b></p>
<ul>
  <li>What happened when you used fewer terms?</li>
  <li>What happened when you increased the number of terms?</li>
  <li>Why does increasing the number of terms change the approximation?</li>
  <li>What does it mean when the result gets closer to a limiting value?</li>
</ul>

In the context of my experiment, convergence is when the result of an operation closes in on a particular number when the number of elements of the operation is increased. <br>
When fewer terms are used, the result is farther from the targeted output. Conversely, the output gets closer to the target when the number of terms is increased. <br>
Increasing the number of terms changes the approximation because it gets getting close the the target output. <br>
When the result gets closer to a limiting value, it means the number of terms is already been increased to the point of achieving the output

<h4>BONUS CHALLENGE</h4>
<p><b>Build Your Own Mini NumPy Numerical Experiment</b></p>

<p>You've completed the three exercises.<br>
Now let's push a little further.</p>

<p>Create your own small numerical experiment using the NumPy concepts from this lesson.</p>

<p><b>Your experiment must include at least five of the following:</b></p>
<ul>
  <li><code>np.array()</code></li>
  <li><code>np.sum()</code></li>
  <li><code>np.mean()</code></li>
  <li><code>np.add()</code></li>
  <li><code>np.subtract()</code></li>
  <li><code>np.multiply()</code></li>
  <li><code>np.divide()</code></li>
  <li><code>np.sqrt()</code></li>
  <li><code>np.arange()</code></li>
  <li><code>np.radians()</code></li>
  <li><code>np.sin()</code></li>
  <li>Element-wise array operations</li>
  <li>A mathematical sequence</li>
</ul>

<br>

<p><b>Your Bonus Challenge Must Have:</b></p>

<p><b>1. A Real-World Scenario</b><br>
Choose an area such as:</p>
<ul>
  <li>Business</li>
  <li>Finance</li>
  <li>Education</li>
  <li>Agriculture</li>
  <li>Healthcare</li>
  <li>Energy</li>
  <li>Engineering</li>
  <li>Sports</li>
  <li>Logistics</li>
  <li>Technology</li>
</ul>

<p><b>2. A Numerical Dataset</b><br>
Create your own NumPy array or sequence.</p>

<p><b>3. At Least Three Calculations</b><br>
Use NumPy to perform at least three meaningful calculations.</p>

<p><b>4. Interpretation</b><br>
Don't just show the numbers.<br>
Explain what the results mean in your chosen scenario.</p>

<p><b>5. Documentation</b><br>
Your notebook should contain:</p>
<ul>
  <li>Project title</li>
  <li>Scenario</li>
  <li>Objective</li>
  <li>Dataset</li>
  <li>Code</li>
  <li>Results</li>
  <li>Interpretation</li>
  <li>Conclusion</li>
</ul>

<b>Yield & Irrigation Efficiency Analysis for Maize Farming</b> <br>
<b>Scenario: Agriculture</b> <br>
You are working as an agricultural analyst for a commercial farm. The management team wants to evaluate daily soil moisture loss across a 10-day period and calculate the required top-up to prevent heat stress in a maize field. <br>

<b>Objective</b> <br>
This project was undertaken to help farm manager get an actionable irrigation schedule to improve maize plant health and avoid over application of irrigation water which leads to leaching of soil nutrients.

<b>Numerical Dataset</b> <br>
days: A sequence of ten numerical data <br>
temperature: An array of ten temperature records in Celsius <br>
moisture: An array of ten moisture levels in percentage <br>

<b>Calculations</b> <br>
Calculate the following: <br>
Daily moisture drop (Evaporation rate) <br>
Moisture deficit relative to the ideal 35% threshold <br>
Irrigation volume <br>
Hottest day



In [52]:
# Code:

# Ten days record:
days = np.arange(1, 11)

# Daily temperature records in Celsius:
temperatures = np.array([28.5, 30.0, 31.5, 33.0, 34.5, 35.0, 34.0, 32.5, 31.0, 29.0])

# Soil moisture levels in % before daily evaporation
moisture_array = np.array([42.0, 40.5, 38.0, 35.0, 31.0, 27.5, 25.0, 24.0, 25.5, 28.0])


# Calculations:
# 1. Daily moisture drop (Daily evaporation rate)
moisture_drop = np.subtract(moisture_array[:-1], moisture_array[1:])


# 2. Moisture deficit relative to the ideal 35%
# The negative values are greater than the ideal moisture level and as such, maize plants don't require any 
# extra irrigation as the soil is already saturated
# Because the negative values require no irrigation, i will make them display as zero

ideal_moisture = 35
deficit = np.subtract(ideal_moisture, moisture_array)
moisture_deficit = np.maximum(0, deficit)


# 3. Irrigation volume
# This is the volume of irrigation water needed based on how far below the ideal moisture level the soil is
# Assuming a 1% drop in moisture requires a 1L of water for irrigation replenishment,
irrigation_volume = np.multiply(moisture_deficit, 1)


# 4. Hottest Day
# The single hottest day is the max data entry for temperture
hottest_day = np.max(temperatures)
# hottest_day_index gives the index of the day with the highest temperature
hottest_day_index = np.argmax(temperatures)


# Result:
print("================================================================================================")
print("                                DATA SUMMARY")
print("================================================================================================")

print("Daily Moisture Drop (%):", moisture_drop)
print()
print("Moisture Deficit below 35% (%):", moisture_deficit)
print()
print("Irrigation Volume Needed (Liters):", irrigation_volume)
print()
print(f"The hottest temperature record was {hottest_day:.0f}°C on day {hottest_day_index + 1}")

print("================================================================================================")



                                DATA SUMMARY
Daily Moisture Drop (%): [ 1.5  2.5  3.   4.   3.5  2.5  1.  -1.5 -2.5]

Moisture Deficit below 35% (%): [ 0.   0.   0.   0.   4.   7.5 10.  11.   9.5  7. ]

Irrigation Volume Needed (Liters): [ 0.   0.   0.   0.   4.   7.5 10.  11.   9.5  7. ]

The hottest temperature record was 35°C on day 6


<b>Interpretation</b> <br>
Using the ten-day data for temperature and soil moisture in a maize farm, this mini-project highlights the drop in soil moisture across the days of moisture data collection, the days where soil moisture fell below the threshold (35%) and showed the hottest day in the ten day window. It also used the data gotten as moisture deficit to infer the approximate volume of irrigation water required by the maize plants using some assumptions (That 1% below moisture threshold correlates with 1L of irrigation water requirement). <br>

<b>Conclusion</b> <br>
With the output of this project, management can get valuable insights as the to the volume of irrigation water required by plants over a heatwave period. This can serve as an actionable irrigation schedule to ensure optimum plant health and prevent heat stress that can inhibit normal physiological and photosynthetic processes.